MREUnityRuntime/MREUnityRuntimeLib/Assets/AssetLoader.cs

```diff
@@ -148,123 +148,135 @@ private async Task LoadAssets(LoadAssets payload, Action onCompleteCallback)
 
 		private async Task<IList<Asset>> LoadAssetsFromGLTF(AssetSource source, Guid containerId, ColliderType colliderType)
 		{
+			WebRequestLoader loader = null;
+			Stream stream = null;
 			IList<Asset> assets = new List<Asset>();
 			DeterministicGuids guidGenerator = new DeterministicGuids(UtilMethods.StringToGuid(
 				$"{containerId}:{source.ParsedUri.AbsoluteUri}"));
 
-			// download file
-			var rootUrl = URIHelper.GetDirectoryName(source.ParsedUri.AbsoluteUri);
-			var loader = new WebRequestLoader(rootUrl);
-			var stream = await loader.LoadStreamAsync(URIHelper.GetFileFromUri(source.ParsedUri));
+			// Wait asynchronously until the load throttler lets us through.
+			using (var scope = await AssetLoadThrottling.AcquireLoadScope())
+			{
+				// download file
+				var rootUrl = URIHelper.GetDirectoryName(source.ParsedUri.AbsoluteUri);
+				loader = new WebRequestLoader(rootUrl);
+				stream = await loader.LoadStreamAsync(URIHelper.GetFileFromUri(source.ParsedUri));
+			}
 
 			// pre-parse glTF document so we can get a scene count
-			// run this in thread
+			// run this on a threadpool thread so that the Unity main thread is not blocked
 			GLTF.Schema.GLTFRoot gltfRoot = null;
-			Thread parseJsonThread = new Thread(() => GLTF.GLTFParser.ParseJson(stream, out gltfRoot));
-			parseJsonThread.Priority = System.Threading.ThreadPriority.Highest;
-			parseJsonThread.Start();
-			RunCoroutineSync(WaitUntilEnum(new WaitUntil(() => !parseJsonThread.IsAlive)));
+			try
+			{
+				await Task.Run(() =>
+				{
+					GLTF.GLTFParser.ParseJson(stream, out gltfRoot);
+				});
+			}
+			catch (Exception e)
+			{
+				Debug.LogError(e);
+			}
 			if (gltfRoot == null)
 			{
 				throw new GLTFLoadException("Failed to parse glTF");
 			}
-			//GLTF.GLTFParser.ParseJson(stream, out GLTF.Schema.GLTFRoot gltfRoot);
 			stream.Position = 0;
 
-			GLTFSceneImporter importer =
-				MREAPI.AppsAPI.GLTFImporterFactory.CreateImporter(gltfRoot, loader, _asyncHelper, stream);
-			importer.SceneParent = MREAPI.AppsAPI.AssetCache.CacheRootGO().transform;
-			importer.Collider = colliderType.ToGLTFColliderType();
-
-			// load textures
-			if (gltfRoot.Textures != null)
+			using (GLTFSceneImporter importer =
+				MREAPI.AppsAPI.GLTFImporterFactory.CreateImporter(gltfRoot, loader, _asyncHelper, stream))
```

Comment:
Oh good catch, I didn't realize glTF loaders were disposable. Are we sure disposing this doesn't destroy the in-memory assets that are in-use?

Reply:
I haven't changed the lifetime of the loader. It was already being disposed, but the call was explicit (line 266 on the left-hand side). This is just a refinement to make the Dispose call implicit, but also makes the lifetime tolerant against exceptions.